### PR TITLE
Template description

### PR DESCRIPTION
**Pull request by bsels**
This pull request adds the functionality to generate descriptions instances for programming languages from a template description.

@niknetniko and @pdawyndt would you carefull look at the `types.json` files foreach programming languages if there are no mistakes in the type names?

### CODE REVIEW

**Review by niknetniko — CHANGES_REQUESTED** [review]
Looks good in general, but I'm thinking the `config.py` file is getting a bit large.

It seems to me the probability that we would want to override the description generation code in a language config is low. Additionally, most of the code seems to be relatively self-contained. As such, I would split the description generation code in another class. If we want to keep the possibility to override them, I would organise it something like this:

In `config.py`:
```python
class Language:
....
    def get_description_generator(params):
         return new DescriptionGenerator(params);
....
```

And then move all the relevant code (e.g. `types`, and the new functions) to a new class `DescriptionGenerator`. If you want to override it then, you can implement a sublcass.

----

(perhaps something for the next meeting)

I'm not 100% sure if we want to support generating the code examples with pygmentize in HTML. I wonder if supporting markdown would be enough for the kind of exercises we envision for TESTed. It just doesn't feel like the responsibility of the judge to me, but I do get why we need it if we want to generate HTML.

File: tested/languages/config.py [hunk 1]
```diff
@@ -636,3 +657,129 @@ def clean_exception_message(self, message: str, namespace: str) -> str:
...
+    def get_type_name(self,
+                      args: TYPE_ARG,
+                      custom_type_map: Optional[Dict[
+                          str, Dict[str, Union[str, Dict[str, str]]]
+                      ]] = None,
+                      bundle: Optional[Bundle] = None,
+                      is_inner: bool = False,
+                      is_html: bool = True,
+                      recursive_call: bool = False) -> str:
+        if bundle is None:
```
[the remark this answers is not in the document]
Comment (bsels) [reply tested/languages/config.py]: It was because if the usage with `functools.partials`, but I have changed and written a wrapper function around it.

**Comment by bsels** [comment]
> I'm not 100% sure if we want to support generating the code examples with pygmentize in HTML. I wonder if supporting markdown would be enough for the kind of exercises we envision for TESTed. It just doesn't feel like the responsibility of the judge to me, but I do get why we need it if we want to generate HTML.

An additional debate if we want to link the filenames inline the statements (instead of listing them above, or probably both) then we also needed pygmentize to generate the html.

**Review by niknetniko — APPROVED** [review]

**Comment by bsels** [comment]
Last commits reuses the description for generating the  HTML for inlining file links and avoids multiple copies of a description_generator.

**Review by niknetniko — APPROVED** [review]

File: tested/languages/generator.py [hunk 2]
```diff
@@ -455,8 +452,12 @@ def get_readable_input(bundle: Bundle,
...
     if not regex.search(text):
         return ExtendedMessage(description=text, format=format_), set()
-    lexer = get_lexer_by_name(format_)
-    generated_html = highlight(text, lexer, _html_formatter)[28:-14]
+
+    if format_ == 'text':
+        generated_html = html.escape(text)
+    else:
+        generator = bundle.lang_config.get_description_generator()
+        generated_html = generator.generate_html_code(text)[28:-14]
```
Comment (pdawyndt) [inline tested/languages/generator.py:460]: Add comments to explain the slice.

File: tested/languages/haskell/types.json [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      "set": "set",
+      "list": "list",
+      "tuple": "tuple"
+    },
+    "nl": {
+      "text": "string",
+      "sequence": "lijst",
+      "set": "verzameling",
+      "list": "lijst",
+      "tuple": "tupel"
```
Comment (pdawyndt) [inline tested/languages/haskell/types.json:55]: I would also use the spelling tuple in Dutch.

File: tested/languages/python/types.json [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      "list": "list",
+      "tuple": "tuple"
+    },
+    "nl": {
+      "text": "string",
+      "sequence": "lijst",
+      "set": "verzameling",
+      "map": "dictionary",
+      "list": "lijst",
+      "tuple": "tupel"
```
Comment (pdawyndt) [inline tested/languages/python/types.json:53]: I would also use the spelling tuple in Dutch.

**Review by pdawyndt — APPROVED** [review]

**Comment by pdawyndt** [comment]
If we use the descriptief generator for the AoC exercises, it would be a good idea to inspect the diffs to see what has changed compared to the descriptions generated by the ad hoc generator. May highlight bugs in any of these generators. Would be nice if we could report on some gains made by using this more robust way of generating descriptions for a given natural language.

**Pull request merged**